### PR TITLE
Average background media color for page background

### LINKED ISSUE #3638: "Page" element can have either color or media applied to it, but not both
## Feature Description

[Figma](https://www.figma.com/file/NHHMwIfxMnz39NxqkrFb7R/Stories---Post-Stable?node-id=2%3A91886)

A user will be able to specify the page background color, or add a page background media element that'd be fullbleed. A user may not specify both color and bg media, as in that case the bg color wouldn't display. 

If the user adds an image, the page background color in the story is automatically set by the user (can be set to be the average pixel color in the image).

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Implementation Brief

**Note**: We will implement this in the old design, not the new one (not yet at least).

Implementation of dominant color will be done using [`colorthief`](https://lokeshdhakar.com/projects/color-thief/), which is already in the project.

For background videos, we take the primary color of the cover image for the video - regardless of whether that cover image is manually set rather than taken from the video itself.

### TODO

- [ ] Remove color panel for background elements if there's a media element
- [ ] Set background element color based on dominant color of media content

### PR DESCRIPTION
**Pull request by miina**
## Summary
Gets the average color of the background media and displays it under the background media as the page background color.
If there is no background media, the panel for Page background color is displayed.

## Relevant Technical Choices
Base color is assigned to each element upon insertion, so it can just be used later.
There are exceptions -- existing media elements and media uploaded by dragging to the editor from somewhere else, for those there’s a workaround that creates a small image out of the screen, gets the average color from there, and assigns the base color together with assigning the background.
The PR changes a bit the logic of how `combineElements` works since the baseColor has to be assigned at the same time to ensure having only one history entry for the action.

## To-do

- [x] Karma

## User-facing changes
- When there's background media, page background color panel is hidden
- In case of background media, the average color of the image/video poster is used as the background color.

## Testing Instructions
1. Add a Page
2. Verify it has the Page background color panel available.
3. Drag an image from library to be the background instead.
4. Verify the panel for Page background color is not visible anymore
5. Verify that the page background color in the front-end is displayed based on the image -- e.g. if you used a pink image as the background media, the page background should have pink color. This can be tested out by disabling cache and setting the mode to slow internet so that while the image is loaded the background color can be seen. Alternatively, the color can be checked from the source code, please ask is there's trouble finding the correct place.
6. Verify that the page background is assigned correctly also in these cases:
- Dragging and dropping existing media to background (both image/video)
- Assigning an existing media as background from the Design Panel (both image/video).

---

Fixes #3638

### CODE REVIEW

File: assets/src/edit-story/components/dropTargets/provider.js [hunk 1]
```diff
@@ -174,26 +175,49 @@ function DropTargetsProvider({ children }) {
...
+          combineArgs.firstElement = {
+            ...firstElement,
+            resource: {
+              ...firstElement.resource,
+              baseColor,
             },
+          };
+          delete combineArgs.firstId;
+        }
+        combineElements(combineArgs);
```
Comment (barklund) [inline assets/src/edit-story/components/dropTargets/provider.js]: Hm, then I would probably rather have, that we **always** pass the whole element as an argument and **never** use `firstId`. Would that be a big change here?

File: assets/src/edit-story/components/dropTargets/provider.js [hunk 2]
```diff
@@ -174,26 +175,49 @@ function DropTargetsProvider({ children }) {
...
-        .filter(({ id }) => !(id in Object.keys(dropTargets)) && id !== selfId)
-        .forEach((el) => {
-          pushTransform(el.id, {
-            dropTargets: {
-              active: false,
-              replacement: null,
+      const firstElement = elements.find(
+        ({ id }) => id === combineArgs.firstId
+      );
+      const finalizeDrop = (baseColor) => {
```
Comment (barklund) [inline assets/src/edit-story/components/dropTargets/provider.js]: Please put in a default value for `baseColor` so it's clear that it is sometimes not set. Just `null` would make sense.

File: assets/src/edit-story/components/library/panes/media/common/innerElement.js [hunk 3]
```diff
@@ -118,7 +132,10 @@ function InnerElement({
       onDragEnd: (e) => {
         e.preventDefault();
         setDraggingResource(null);
-        handleDrop(resource);
+        handleDrop({
+          ...resource,
+          baseColor: mediaBaseColor.current,
```
Comment (barklund) [inline assets/src/edit-story/components/library/panes/media/common/innerElement.js:137]: What if it hasn't been calculated yet? Is that even an option?

File: assets/src/edit-story/components/panels/index.js [hunk 4]
```diff
@@ -119,8 +117,11 @@ export function getPanels(elements, options = {}) {
         Panel: ImageAccessibilityPanel,
       });
     }
-    // Always display Presets as the first panel for background.
-    panels.unshift({ type: STYLE_PRESETS, Panel: ColorPresetPanel });
+    if (!isBackgroundMedia) {
```
Comment (barklund) [inline assets/src/edit-story/components/panels/index.js]: `else`?

File: assets/src/edit-story/utils/getMediaBaseColor.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  position: 'fixed',
+  contain: 'layout paint',
+  top: '-9999px',
+  left: '-9999px',
+  zIndex: -1,
+};
+
+const BASE_COLOR_NODE = '__WEB_STORIES_BASE_COLOR__';
+
+export function getMediaBaseColor(type, resource, onBaseColor) {
```
Comment (barklund) [inline assets/src/edit-story/utils/getMediaBaseColor.js]: Doesn't the `resource` have the `type` as well?

File: assets/src/edit-story/utils/getMediaBaseColor.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  for (const k in styles) {
+    if (Object.prototype.hasOwnProperty.call(styles, k)) {
+      const v = styles[k];
+      if (v === null) {
+        node.style[k] = '';
+      } else {
+        node.style[k] = v;
+      }
+    }
+  }
```
Comment (barklund) [inline assets/src/edit-story/utils/getMediaBaseColor.js]: This could be done simply as this - probably doesn't need a helper function: `Object.assign(node.style, styles);`.

**Review by barklund — COMMENTED** [review]
Overall it looks good - the direction seems sound and the result looks good.

I'm always hesitant with logic appearing in multiple places (here extracting `baseColor` happening in 2 very different but similar places), but I guess it's necessary in this instance for backwards compatibility.

File: assets/src/edit-story/components/dropTargets/provider.js [hunk 7]
```diff
@@ -174,26 +175,49 @@ function DropTargetsProvider({ children }) {
...
-      const { onDropHandler } = getDefinitionForType(resource.type);
-      if (onDropHandler) {
-        onDropHandler(activeDropTargetId);
+        const { onDropHandler } = getDefinitionForType(resource.type);
+        if (onDropHandler) {
+          onDropHandler(activeDropTargetId);
+        }
+      };
+      // Skip if we already have the color.
+      if (firstElement.baseColor) {
```
Comment (barklund) [inline assets/src/edit-story/components/dropTargets/provider.js]: This by the way fails (`Cannot read property 'baseColor' of undefined`), when you drop an element directly from the library to the background drop target.

File: assets/src/edit-story/components/library/panes/media/common/innerElement.js (hunk 3 above)
Comment (miina) [reply assets/src/edit-story/components/library/panes/media/common/innerElement.js:137]: If for some reason it's not calculated, the logic in the drop handler will still take care of it.

**Comment by miina** [comment]
> I'm always hesitant with logic appearing in multiple places (here extracting baseColor happening in 2 very different but similar places), but I guess it's necessary in this instance for backward compatibility.

Yep, was trying to think of a better alternative but didn't have one yet. One option could be not assigning the base color in the media library at all and only doing it when combining two elements -- would have it in one place (leaving only the logic for the current backward compatibility) but would make combining a bit slower (since the base color always needs to be extracted first). In this case, perhaps it's OK to still leave it in two places.

File: assets/src/edit-story/components/dropTargets/provider.js (hunk 1 above)
Comment (miina) [reply assets/src/edit-story/components/dropTargets/provider.js]: Shouldn't be a big change, will check.

File: assets/src/edit-story/utils/getMediaBaseColor.js [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import ColorThief from 'colorthief';
+const thief = new ColorThief();
+
+const STYLES = {
```
Comment (barklund) [inline assets/src/edit-story/utils/getMediaBaseColor.js:23]: `const HARRY` 😎

File: assets/src/edit-story/app/story/useStoryReducer/reducers/combineElements.js [hunk 9]
```diff
@@ -46,15 +46,15 @@ import objectWithout from '../../../../utils/objectWithout';
...
-function combineElements(state, { firstId, firstElement, secondId }) {
-  if ((!firstId && !firstElement) || !secondId) {
+function combineElements(state, { firstElement, secondId }) {
+  if (!firstElement || !secondId) {
     return state;
   }
+  const firstId = firstElement.id;
 
   const pageIndex = state.pages.findIndex(({ id }) => id === state.current);
   const page = state.pages[pageIndex];
```
Comment (barklund) [inline assets/src/edit-story/app/story/useStoryReducer/reducers/combineElements.js:61]: So, two lines below here, shouldn't `element` always just be `firstElement`?

Otherwise, there's easily an issue where the passed element has an ID of an existing element, and then the passed element is ignored? That was the whole purpose of either using ID or element before, but rewriting to only using full element now.

The ID should be used to find which old element is to be removed if any, but nothing more than that.

(can't comment on lines too far from lines changed in this PR, unfortunately).

File: assets/src/edit-story/app/story/useStoryReducer/test/combineElements.js [hunk 10]
```diff
@@ -47,40 +50,55 @@ describe('combineElements', () => {
     const initial = restore(getDefaultState1());
 
     // Combine non-existing element abc into 789
-    const result = combineElements({ firstId: 'abc', secondId: '789' });
+    const result = combineElements({
+      firstElement: { id: 'abc', type: 'image' },
+      secondId: '789',
+    });
```
Comment (barklund) [inline assets/src/edit-story/app/story/useStoryReducer/test/combineElements.js]: I don't think this test makes sense anymore. This should actually do a merge and overwrite with new props as long as the target ID exists. Doesn't it do that?

File: assets/src/edit-story/components/dropTargets/karma/order.karma.js [hunk 11]
```diff
@@ -54,14 +59,13 @@ describe('Drop-Target order', () => {
...
     ).node;
-
     await fixture.events.mouse.seq(({ moveRel, moveBy, down, up }) => [
       moveRel(replacementImageFrame, 10, 10),
       down(),
       moveBy(0, -20),
       up(),
     ]);
-
+    await waitForElementToBeRemoved(replacementImageFrame);
```
Comment (barklund) [inline assets/src/edit-story/components/dropTargets/karma/order.karma.js:68]: This is added because? Flaky tests or?

File: assets/src/edit-story/components/dropTargets/provider.js [hunk 12]
```diff
@@ -174,26 +176,45 @@ function DropTargetsProvider({ children }) {
...
+              baseColor,
             },
+          };
+        }
+        combineElements(combineArgs);
+
+        // Reset styles on visible elements
+        elements
+          .filter(
+            ({ id }) => !(id in Object.keys(dropTargets)) && id !== selfId
```
Comment (barklund) [inline assets/src/edit-story/components/dropTargets/provider.js]: Wait, that's not how [`in`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/in) works.

```js
console.log(1 in [1]); // false
```

Either use `id in dropTargets`, `dropTargets[id]`, or `Object.keys(dropTarget).includes(id)`.

This appears to be a weird hybrid, that doesn't make any sense!

File: assets/src/edit-story/components/dropTargets/provider.js [hunk 13]
```diff
@@ -174,26 +176,45 @@ function DropTargetsProvider({ children }) {
...
-              replacement: null,
+      const finalizeDrop = (baseColor = null) => {
+        if (baseColor) {
+          combineArgs.firstElement = {
+            ...combineArgs.firstElement,
+            resource: {
+              ...combineArgs.firstElement.resource,
+              baseColor,
             },
+          };
```
Comment (barklund) [inline assets/src/edit-story/components/dropTargets/provider.js]: You could probably also just do `combineArgs.firstElement.resource.baseColor = baseColor`, as `resource` has to exist on the existing `firstElement`, right?

File: assets/src/edit-story/components/panels/index.js [hunk 14]
```diff
@@ -90,15 +90,13 @@ export function getPanels(elements, options = {}) {
   }
 
   const isBackground = elements.length === 1 && elements[0].isBackground;
+  const isBackgroundMedia = isBackground && !elements[0].isDefaultBackground;
```
Comment (barklund) [inline assets/src/edit-story/components/panels/index.js]: This could probably be moved to just before usage and then the addition of `isBackground &&` appears irrelevant.

File: assets/src/edit-story/components/panels/sizePosition/panel.js [hunk 15]
```diff
@@ -131,10 +132,30 @@ function SizePositionPanel({
...
+          },
+          secondId: currentBackgroundId,
+        });
+      }
+    };
+    if (selectedElements[0].resource?.baseColor) {
+      setBackground();
+    } else {
+      getMediaBaseColor(selectedElements[0].resource, setBackground);
+    }
```
Comment (barklund) [inline assets/src/edit-story/components/panels/sizePosition/panel.js]: If there's no resource, the `else` branch will be used, which will probably fail. So the  `?` in `selectedElements[0].resource?.baseColor` should probably be removed or handled separately.

**Review by barklund — COMMENTED** [review]
Could we try to add a karma test for this one? I'd love to see that it works all the way from adding an image to the resulting AMP output.

File: assets/src/edit-story/components/dropTargets/provider.js (hunk 12 above)
Comment (miina) [reply assets/src/edit-story/components/dropTargets/provider.js]: Hmm, I actually didn't look at this part at all, left it as it was before, will check...

File: assets/src/edit-story/components/dropTargets/karma/order.karma.js (hunk 11 above)
Comment (miina) [reply assets/src/edit-story/components/dropTargets/karma/order.karma.js:68]: Yes, to ensure waiting until the original image has been removed and the dropping has happened.

File: assets/src/edit-story/app/story/useStoryReducer/test/combineElements.js (hunk 10 above)
Comment (miina) [reply assets/src/edit-story/app/story/useStoryReducer/test/combineElements.js]: Hmm, why so? I assumed that if there's an incorrect ID then this means that the origin element is incorrectly passed and nothing should happen. Because for drop targeting, the origin element either should not have an ID (is coming from the library for example) or should have an existing ID (an existing media is dropped into something). There shouldn't be a case where an element with incorrect ID is dropped. Thoughts?

Comment (miina) [reply assets/src/edit-story/app/story/useStoryReducer/test/combineElements.js]: Although maybe that assumption was just in my head and it actually does already work as you're saying it should.

Will refactor a bit assuming that the ID can be passed incorrectly as well and then just the props should be taken.

File: assets/src/edit-story/app/story/useStoryReducer/reducers/combineElements.js (hunk 9 above)
Comment (miina) [reply assets/src/edit-story/app/story/useStoryReducer/reducers/combineElements.js:61]: OK, changing -- I assumed for some reason that incorrect ID from the first element should abort the mission, however, apparently it's not doing it anyway currently either.

File: assets/src/edit-story/components/dropTargets/provider.js (hunk 12 above)
Comment (miina) [reply assets/src/edit-story/components/dropTargets/provider.js]: Ok, looks like since this condition has always been set to `true` so far then it has always reset the style for all elements, including irrelevant elements. Changed in a064519.

**Comment by miina** [comment]
> Could we try to add a karma test for this one? I'd love to see that it works all the way from adding an image to the resulting AMP output.

Hm, since the Karma tests are taking a long time already now to run, I was actually purposefully adding it to already existing Karma tests since setting the color is one part of setting the background, even though we added it separately. Currently, there are two tests where the added `baseColor` is being checked + separate unit tests to ensure that if it's present in the props, it will be added to the HTML properly as well.

Let me know if you still think it's worthwhile trying to add a separate Karma test focused on this one feature specifically.

**Review by barklund — APPROVED** [review]
Works great!

**Comment by csossi** [comment]
Verified in QA

example (while loading)
